store/mongo/datastore_mongo.go

```diff
@@ -360,38 +422,110 @@ func (db *DataStoreMongo) GetSessionRecording(ctx context.Context,
+		control := controlReader.Pop()
+		if control == nil {
+			l.Debug("WriteSessionRecords: no more control " +
+				"messages, flushing the recording upstream.")
+			//no more control messages, we send the whole recording
+			var err error = nil
+			var n int
+			for err != io.EOF {
+				n, err = recordingReader.Read(recordingBuffer)
+				if n > 0 {
+					//nolint:errcheck
+					sendRecordingMessage(recordingBuffer[:n], sessionID, w)
+					recordingBytesSent += n
+				}
+			}
+			break
+		} else {
+			if recordingBytesSent > control.Offset {
+				//this should never happen, we missed
+				//the control message, no idea what to do
+				l.Errorf("WriteSessionRecords: recordingBytesSent > control.Offset")
 				break
 			}
-			_, e = w.Write(output[:n])
-			if e != nil {
-				err = e
+
+			bytesUntilControlMessage := control.Offset - recordingBytesSent
+			l.Debugf("(1) WriteSessionRecords: control.Offset:%d"+
+				" recordingBytesSent:%d "+
+				" bytesUntilControlMessage:%d (recordingBuffer.len=%d) "+
+				"reading up to %d bytes of recording and sending.",
+				control.Offset,
+				recordingBytesSent,
+				bytesUntilControlMessage,
+				len(recordingBuffer),
+				control.Offset-recordingBytesSent)
+			//it is possible that the recording is larger than one recordingBuffer,
+			//we need to send until we have the control.Offset in the buffer
+			for bytesUntilControlMessage > len(recordingBuffer) {
+				n, e := recordingReader.Read(recordingBuffer)
+				if n > 0 {
+					//error handling: see below
+					//nolint:errcheck
+					sendRecordingMessage(recordingBuffer[:n], sessionID, w)
+					recordingBytesSent += n
+					bytesUntilControlMessage = control.Offset -
+						recordingBytesSent
+				}
+				if e != nil || n == 0 {
+					break
+				}
+			}
+
+			bytesUntilControlMessage = control.Offset - recordingBytesSent
+			l.Debugf("(2) WriteSessionRecords: control.Offset:%d"+
+				" recordingBytesSent:%d "+
+				" bytesUntilControlMessage:%d (recordingBuffer.len=%d) "+
+				"reading up to %d bytes of recording and sending.",
+				control.Offset,
+				recordingBytesSent,
+				bytesUntilControlMessage,
+				len(recordingBuffer),
+				bytesUntilControlMessage)
+			//this means that the control offset is in the future
+			//part of the recording buffer
+			//we can send up to control.Offset-recordingBytesSent
+			//bytes and then send the control message
+			n, e := recordingReader.Read(recordingBuffer[:bytesUntilControlMessage])
+			l.Debugf("recordingReader.Read(len=%d)=%d,%+v",
+				bytesUntilControlMessage, n, e)
+			if n > 0 {
+				//lets just ignore the error, in so many places
+				//in theory it may go wrong,
+				//and we do not know what to do in most of them
+				//otherwise than pretend it passed
+				//nolint:errcheck
+				sendRecordingMessage(recordingBuffer[:n], sessionID, w)
+				recordingBytesSent += n
 			}
+			l.Debugf("WriteSessionRecords: sending %+v.", *control)
+			//nolint:errcheck
+			sendControlMessage(*control, sessionID, w)
```

Comment:
I don't really understand this part, what if the above loop successfully copies the entire record up to the control message?

Reply:
then it sends the control message, and continues.

Reply:
that is, the idea here is:
1. get the control message, send the recording data up to it
2. send the control message, ->1.
3. if there are no more control messages just flush the whole recording.